### PR TITLE
Pub (Dart) Support: Initial attempt to add minimal support for Dart p…

### PR DESCRIPTION
**Pull request by moo24**
…ackages managed via https://pub.dev/. See https://github.com/heremaps/oss-review-toolkit/issues/757.
I am not sure how to add fully support for flutter, as a scan for Android and iOS packages (Graddle, Cocoapods) might be required for each package where "sdks" include flutter. As "flutter packages get" will also install native packages using these package managers. However, a working pub dependency resolution will be the first step towards full Flutter support.
Please review and provide feedback. Thanks.

---
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/1539)

### CODE REVIEW

File: analyzer/src/funTest/assets/projects/synthetic/pub-expected-output.yml [hunk 1]
```diff
@@ -0,0 +1,1620 @@
...
+  vcs_processed:
+    type: "git"
+    url: "ssh://git@github.com/dart-lang/http.git"
+    revision: "683c29c101c63e79fd9eced14e37a8b4dea23ad4"
+    path: ""
+  homepage_url: "https://github.com/dart-lang/http"
+  scopes:
+  - name: "dependencies"
+    dependencies:
+    - id: "Pup:async:async:2.2.0"
```
[the remark this answers is not in the document]
Comment (moo24) [reply analyzer/src/funTest/assets/projects/synthetic/pub-expected-output.yml]: The dependency tree for Dart looks fine now. For Flutter, all Android packages used in flutter packages are found and added to list of packages. 

However, when I try to add them to the dependency tree, I get the error that the package an unknown package. (IllegalArgumentException: The following references do not actually refer\
    \ to packages: [Identifier(type=Maven, namespace=com.android.tools, name=annotations,\
    \ version=26.3.2)]."
  severity: "ERROR")  

But the package com.android.tools.annotations is listed in packages (see expected output  yml file. (I generated it with disabling the building of the dependency tree for Flutter packages via adding 'false &&' in the if in line 216. I am wondering a bit what´s causing the error. 

Another issue of adding them directly to the dart tree, we will mix the 'scopes'. Maybe it is better / easier to add just the additional scopes. However, then we loose the information which flutter package depends on which android / iOS packages - which might cause issues during debugging of Scan Results.

Maybe you will find some time to review and then we can have a quick discussion.

File: .gitmodules [hunk 2]
```diff
@@ -34,3 +34,6 @@
 	path = analyzer/src/funTest/assets/projects/external/sbt-multi-project-example
 	url = https://github.com/pbassiner/sbt-multi-project-example.git
 	ignore = untracked
+[submodule "http"]
+	path = analyzer/src/funTest/assets/projects/external/http
```
Comment (mnonnenmacher) [inline .gitmodules]: It would be helpful to use "dart-http" for the submodule name and path for clarity.

File: analyzer/src/funTest/assets/projects/external/http-expected-output.yml [hunk 3]
```diff
@@ -0,0 +1,1453 @@
...
+  declared_licenses: []
+  declared_licenses_processed: {}
+  vcs:
+    type: "git"
+    url: "https://github.com/dart-lang/http.git"
+    revision: ""
+    path: ""
+  vcs_processed:
+    type: "git"
+    url: "ssh://git@github.com/dart-lang/http.git"
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/assets/projects/external/http-expected-output.yml]: This should also be the https URL.

File: analyzer/src/funTest/assets/projects/synthetic/pub/no-lockfile/.dart_tool/pub/bin/sdk-version [hunk 4]
```diff
@@ -0,0 +1 @@
+2.3.2-dev.0.0.flutter-e3edfd36b2
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/assets/projects/synthetic/pub/no-lockfile/.dart_tool/pub/bin/sdk-version]: I guess this file should not be part of the commit?

File: analyzer/src/funTest/assets/projects/synthetic/pub/no-lockfile/.packages [hunk 5]
```diff
@@ -0,0 +1,58 @@
+# Generated by pub on 2019-06-13 14:21:50.038631.
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/assets/projects/synthetic/pub/no-lockfile/.packages]: This file should be removed from the commit. Also please add a .gitignore to this directory including this file and others that are generated when running the analyzer.

File: analyzer/src/funTest/kotlin/PubTest.kt [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+import com.here.ort.analyzer.managers.Pub
+import com.here.ort.downloader.VersionControlSystem
+import com.here.ort.model.config.AnalyzerConfiguration
+import com.here.ort.model.yamlMapper
+import com.here.ort.utils.normalizeVcsUrl
+import com.here.ort.utils.test.DEFAULT_ANALYZER_CONFIGURATION
+import com.here.ort.utils.test.DEFAULT_REPOSITORY_CONFIGURATION
+import com.here.ort.utils.test.USER_DIR
+import com.here.ort.utils.test.patchExpectedResult
+import io.kotlintest.shouldBe
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/kotlin/PubTest.kt]: Add a blank line before this one.

File: analyzer/src/funTest/kotlin/PubTest.kt [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        "Pub" should {
+            "resolve dart http dependencies correctly" {
+                val workingDir = File(projectsDirExternal, "http")
+                val packageFile = File(workingDir, "pubspec.yaml")
+                var expectedResultFile = File(projectsDirExternal, "http-expected-output.yml");
+
+                val result = createPubForExternal().resolveDependencies(listOf(packageFile))[packageFile]
+                result!!.errors.size shouldBe 0
+
+                //expectedResultFile.writeText(yamlMapper.writeValueAsString(result));
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/kotlin/PubTest.kt]: Remove commented code.

File: analyzer/src/funTest/kotlin/PubTest.kt [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    init {
+
+        "Pub" should {
+            "resolve dart http dependencies correctly" {
+                val workingDir = File(projectsDirExternal, "http")
+                val packageFile = File(workingDir, "pubspec.yaml")
+                var expectedResultFile = File(projectsDirExternal, "http-expected-output.yml");
+
+                val result = createPubForExternal().resolveDependencies(listOf(packageFile))[packageFile]
+                result!!.errors.size shouldBe 0
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/kotlin/PubTest.kt]: This check is implicitly done by the comparison below.

File: analyzer/src/funTest/kotlin/PubTest.kt [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+                var expectedResultFile = File(projectsDir.parentFile, "pub-expected-output-project-with-flutter.yml");
+
+                val result = createPub().resolveDependencies(listOf(packageFile))[packageFile]
+
+                if(!result!!.errors.isEmpty()) {
+                    print(result.errors.get(0).message);
+                }
+                //result.errors.size shouldBe 0
+
+                //expectedResultFile.writeText(yamlMapper.writeValueAsString(result));
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/kotlin/PubTest.kt]: Remove commented code.

File: analyzer/src/funTest/kotlin/PubTest.kt [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+                val workingDir = File(projectsDir, "project-with-flutter")
+                val packageFile = File(workingDir, "pubspec.yaml")
+                var expectedResultFile = File(projectsDir.parentFile, "pub-expected-output-project-with-flutter.yml");
+
+                val result = createPub().resolveDependencies(listOf(packageFile))[packageFile]
+
+                if(!result!!.errors.isEmpty()) {
+                    print(result.errors.get(0).message);
+                }
+                //result.errors.size shouldBe 0
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/kotlin/PubTest.kt]: Remove debugging code, also the check is implicitly done by the comparison below.

File: analyzer/src/funTest/kotlin/PubTest.kt [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+                var expectedResultFile = File(projectsDir.parentFile, "pub-expected-output-any-version.yml");
+
+                val result = createPub().resolveDependencies(listOf(packageFile))[packageFile]
+
+                if(!result!!.errors.isEmpty()) {
+                    print(result.errors.get(0).message);
+                }
+                result.errors.size shouldBe 0
+
+                //expectedResultFile.writeText(yamlMapper.writeValueAsString(result));
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/kotlin/PubTest.kt]: Remove commented code.

File: analyzer/src/funTest/kotlin/PubTest.kt [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+                val workingDir = File(projectsDir, "any-version")
+                val packageFile = File(workingDir, "pubspec.yaml")
+                var expectedResultFile = File(projectsDir.parentFile, "pub-expected-output-any-version.yml");
+
+                val result = createPub().resolveDependencies(listOf(packageFile))[packageFile]
+
+                if(!result!!.errors.isEmpty()) {
+                    print(result.errors.get(0).message);
+                }
+                result.errors.size shouldBe 0
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/kotlin/PubTest.kt]: This check is implicitly done by the comparison below. If you remove the check the diff show of the comparison shows the error message.

File: analyzer/src/funTest/kotlin/PubTest.kt [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+                    expectedResultFile,
+                    custom = Pair("pub-project", "pub-${workingDir.name}"),
+                    definitionFilePath = "$vcsPath/pubspec.yaml",
+                    url = normalizeVcsUrl(vcsUrl),
+                    revision = vcsRevision,
+                    path = vcsPath
+                )
+
+                yamlMapper.writeValueAsString(result) shouldBe expectedResult
+
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/kotlin/PubTest.kt]: Remove empty line.

File: analyzer/src/funTest/kotlin/PubTest.kt [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/kotlin/PubTest.kt]: Remove empty lines.

File: analyzer/src/funTest/kotlin/PubTest.kt [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+                val result = createPub().resolveDependencies(listOf(packageFile))[packageFile]
+
+                result shouldNotBe null
+                result!!.project.definitionFilePath shouldBe
+                        "analyzer/src/funTest/assets/projects/synthetic/pub/no-lockfile/pubspec.yaml"
+                result.packages.size shouldBe 0
+                result.errors.size shouldBe 1
+                result.errors.first().message should startWith("IllegalArgumentException: No lock file found in")
+            }
+
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/kotlin/PubTest.kt]: Remove empty line.

File: analyzer/src/test/kotlin/PackageManagerTest.kt [hunk 16]
```diff
@@ -57,6 +57,7 @@ class PackageManagerTest : WordSpec({
             managedFilesByName["SBT"] shouldBe listOf(File(projectDir, "build.sbt"))
             managedFilesByName["Stack"] shouldBe listOf(File(projectDir, "stack.yaml"))
             managedFilesByName["Yarn"] shouldBe listOf(File(projectDir, "package.json"))
+            managedFilesByName["Pub"] shouldBe listOf(File(projectDir, "pubspec.yaml"))
```
Comment (mnonnenmacher) [inline analyzer/src/test/kotlin/PackageManagerTest.kt]: Put this in alphabetic order.

File: model/src/main/kotlin/ProjectAnalyzerResult.kt [hunk 17]
```diff
@@ -89,4 +89,27 @@ data class ProjectAnalyzerResult(
...
+
+        for (scope in project.scopes) {
+            if(scope.name == filterByScopeName) {
+                for (dependency in scope.dependencies) {
+                    addPackage(dependency)
+                }
+            }
+        }
+
+        return collectedPackages.distinct();
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/ProjectAnalyzerResult.kt]: You can simplify this function as follows:

```kotlin
        val scope = project.scopes.find { it.name == filterByScopeName } ?: return emptyList()
        return packages.filter { it.pkg.id in scope }
```

File: model/src/main/kotlin/ProjectAnalyzerResult.kt [hunk 18]
```diff
@@ -89,4 +89,27 @@ data class ProjectAnalyzerResult(
             }
         }
     }
+
+    fun collectPackages(filterByScopeName: String): List<CuratedPackage> {
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/ProjectAnalyzerResult.kt]: Consider renaming this function to `collectPackagesByScope(scopeName: String)`.

File: analyzer/src/main/kotlin/managers/Pub.kt [hunk 19]
```diff
@@ -0,0 +1,538 @@
...
+import com.here.ort.model.Project
+import com.here.ort.model.ProjectAnalyzerResult
+import com.here.ort.model.RemoteArtifact
+import com.here.ort.model.Scope
+import com.here.ort.model.yamlMapper
+import com.here.ort.model.VcsInfo
+import com.here.ort.model.config.AnalyzerConfiguration
+import com.here.ort.model.config.RepositoryConfiguration
+import com.here.ort.model.OrtIssue
+
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Pub.kt]: Remove empty line.

File: analyzer/src/main/kotlin/managers/Pub.kt [hunk 20]
```diff
@@ -0,0 +1,538 @@
...
+import com.here.ort.utils.CommandLineTool
+import com.here.ort.utils.ProcessCapture
+import com.here.ort.utils.log
+import com.here.ort.utils.showStackTrace
+import com.here.ort.utils.textValueOrEmpty
+import com.here.ort.utils.collectMessagesAsString
+import com.here.ort.utils.getUserHomeDirectory
+import com.vdurmont.semver4j.Requirement
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Pub.kt]: Add an empty line above this one.

File: analyzer/src/main/kotlin/managers/Pub.kt [hunk 21]
```diff
@@ -0,0 +1,538 @@
...
+) : PackageManager(name, analyzerRoot, analyzerConfig, repoConfig), CommandLineTool {
+    class Factory : AbstractPackageManagerFactory<Pub>("Pub") {
+        override val globsForDefinitionFiles = listOf("pubspec.yaml")
+
+        override fun create(
+            analyzerRoot: File,
+            analyzerConfig: AnalyzerConfiguration,
+            repoConfig: RepositoryConfiguration
+        ) =
+            Pub(managerName, analyzerRoot, analyzerConfig, repoConfig)
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Pub.kt]: Move this to the line above.

File: analyzer/src/main/kotlin/managers/Pub.kt [hunk 22]
```diff
@@ -0,0 +1,538 @@
...
+    }
+
+    override fun getVersionRequirement(): Requirement = Requirement.buildIvy("[2.2,)")
+
+    override fun beforeResolution(definitionFiles: List<File>) {
+
+    }
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Pub.kt]: No need to override this function.

File: analyzer/src/main/kotlin/managers/Pub.kt [hunk 23]
```diff
@@ -0,0 +1,538 @@
...
+
+    /**
+     * A workspace reader that is backed by the local Gradle artifact cache.
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Pub.kt]: Adapt this copied comment to this class.

File: analyzer/src/main/kotlin/managers/Pub.kt [hunk 24]
```diff
@@ -0,0 +1,538 @@
...
+     */
+    private class PubCacheReader {
+        // TODO: Check if we can support a second .pub-cache
+        // If there is a .pub-cache directory in Flutter root,
+        // flutter packages get will use that one instead of the .pub-cache directory in user home
+        // Currently we do not support usage of two different pub-cache directories.
+        // Therefore you need to delete the .pub-cache directory in your flutter root folder
+        // to force flutter to use the default one
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Pub.kt]: Please use the max allowed line length of 120 for this comment, to make it take less lines.

File: analyzer/src/funTest/kotlin/PubTest.kt (hunk 12 above)
Comment (moo24) [reply analyzer/src/funTest/kotlin/PubTest.kt]: done

File: analyzer/src/funTest/kotlin/PubTest.kt (hunk 11 above)
Comment (moo24) [reply analyzer/src/funTest/kotlin/PubTest.kt]: done

File: analyzer/src/funTest/kotlin/PubTest.kt (hunk 14 above)
Comment (moo24) [reply analyzer/src/funTest/kotlin/PubTest.kt]: done

File: analyzer/src/funTest/kotlin/PubTest.kt (hunk 15 above)
Comment (moo24) [reply analyzer/src/funTest/kotlin/PubTest.kt]: done

File: analyzer/src/test/kotlin/PackageManagerTest.kt (hunk 16 above)
Comment (moo24) [reply analyzer/src/test/kotlin/PackageManagerTest.kt]: done

File: model/src/main/kotlin/ProjectAnalyzerResult.kt (hunk 18 above)
Comment (moo24) [reply model/src/main/kotlin/ProjectAnalyzerResult.kt]: done

File: model/src/main/kotlin/ProjectAnalyzerResult.kt (hunk 17 above)
Comment (moo24) [reply model/src/main/kotlin/ProjectAnalyzerResult.kt]: done

File: analyzer/src/funTest/kotlin/PubTest.kt (hunk 7 above)
Comment (moo24) [reply analyzer/src/funTest/kotlin/PubTest.kt]: done

File: analyzer/src/funTest/kotlin/PubTest.kt (hunk 8 above)
Comment (moo24) [reply analyzer/src/funTest/kotlin/PubTest.kt]: done

File: analyzer/src/funTest/kotlin/PubTest.kt (hunk 6 above)
Comment (moo24) [reply analyzer/src/funTest/kotlin/PubTest.kt]: done

File: analyzer/src/funTest/assets/projects/external/http-expected-output.yml (hunk 3 above)
Comment (moo24) [reply analyzer/src/funTest/assets/projects/external/http-expected-output.yml]: changed

File: .gitmodules (hunk 2 above)
Comment (moo24) [reply .gitmodules]: changed

File: analyzer/src/funTest/assets/projects/synthetic/pub/no-lockfile/.packages (hunk 5 above)
Comment (moo24) [reply analyzer/src/funTest/assets/projects/synthetic/pub/no-lockfile/.packages]: done

File: analyzer/src/funTest/assets/projects/synthetic/pub/no-lockfile/.dart_tool/pub/bin/sdk-version (hunk 4 above)
Comment (moo24) [reply analyzer/src/funTest/assets/projects/synthetic/pub/no-lockfile/.dart_tool/pub/bin/sdk-version]: removed

File: analyzer/src/main/kotlin/managers/Pub.kt [hunk 25]
```diff
@@ -0,0 +1,533 @@
...
+    private class PubCacheReader {
+        // If there is a .pub-cache directory in Flutter root, 'flutter packages get'
+        // will use that one instead of the .pub-cache directory in user home
+        // Currently we do not support usage of two different pub-cache directories.
+        // Therefore you need to delete the .pub-cache directory in your flutter root folder
+        // to force flutter to use the default one
+        private val pubCacheRoot = getUserHomeDirectory().resolve(".pub-cache/")
+
+        fun findFile(packageInfo: JsonNode, fileName: String): File? {
+            val artifactRootDir = findProjectRoot(packageInfo);
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Pub.kt]: Please remove all semicolons in this file.

File: analyzer/src/main/kotlin/managers/Pub.kt [hunk 26]
```diff
@@ -0,0 +1,533 @@
...
+
+            // try to locate file directly
+            val file = File(
+                artifactRootDir,
+                fileName
+            );
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Pub.kt]: This would fit in one line.

File: analyzer/src/main/kotlin/managers/Pub.kt [hunk 27]
```diff
@@ -0,0 +1,533 @@
...
+            var path : String;
+
+            val packageVersion = packageInfo["version"].textValueOrEmpty();
+            val type = packageInfo["source"].textValueOrEmpty();
+            val description = packageInfo["description"];
+            val packageName = description["name"].textValueOrEmpty();
+            val url = description["url"].textValueOrEmpty();
+            val resolvedRef = packageInfo["resolved-ref"].textValueOrEmpty();
+
+            if(type == "hosted" && url.isNotEmpty()){
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Pub.kt]: Missing space after if. Best use the IntelliJ formatter to format this file, this should remove all similar formatting issues.

File: analyzer/src/main/kotlin/managers/Pub.kt [hunk 28]
```diff
@@ -0,0 +1,533 @@
...
+                // not supported type
+                path = "/";
+
+                log.error { "Could not find projectRoot of '$packageName'" }
+            }
+
+            val artifactRootDir = File(
+                pubCacheRoot,
+                path
+            )
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Pub.kt]: This would fit in one line.

File: analyzer/src/main/kotlin/managers/Pub.kt [hunk 29]
```diff
@@ -0,0 +1,533 @@
...
+
+        val manifest = yamlMapper.readTree(definitionFile)
+        val hasDependencies = manifest.fields().asSequence().any { (key, value) ->
+            key.startsWith("dependencies") && value.count() > 0
+        }
+
+        val (packages, scopes) = if (hasDependencies) {
+            installDependencies(workingDir)
+
+            log.info { "Reading $PUB_LOCK_FILE file in $workingDir " }
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Pub.kt]: End the message with a dot and remove the trailing slashes in this log message and the three below.

File: analyzer/src/main/kotlin/managers/Pub.kt [hunk 30]
```diff
@@ -0,0 +1,533 @@
...
+        } else {
+            Pair(emptyMap(), sortedSetOf())
+        }
+
+        log.info { "Reading ${definitionFile.name} file in $workingDir " }
+
+        val project = parseProject(definitionFile, scopes)
+
+        return ProjectAnalyzerResult(project, packages.values.map { it.toCuratedPackage() }.toSortedSet())
+
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Pub.kt]: Remove empty line.

File: analyzer/src/main/kotlin/managers/Pub.kt [hunk 31]
```diff
@@ -0,0 +1,533 @@
...
+    }
+
+    private fun parseScope(
+        scopeName: String, manifest: JsonNode, lockFile: JsonNode, packages: Map<Identifier, Package>
+    ): Scope {
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Pub.kt]: If the function signature does not fit in one line, put each argument in a separate line, following the Kotlin style guide.

File: analyzer/src/main/kotlin/managers/Pub.kt [hunk 32]
```diff
@@ -0,0 +1,533 @@
...
+            }
+        } ?: VcsInfo.EMPTY
+    }
+
+    override fun command(workingDir: File?) = "pub"
+
+    private fun commandFlutter() = "flutter packages"
+
+    /*override fun run(workingDir: File?, vararg args: String) =
+        ProcessCapture(workingDir, *command(workingDir).split(" ").toTypedArray(), *args).requireSuccess()*/
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Pub.kt]: Remove commented code.
